data/transactions/logic/eval.go

```diff
@@ -678,33 +678,46 @@ func (st StackType) Typed() bool {
 	return false
 }
 
-// PanicError wraps a recover() catching a panic()
-type PanicError struct {
+// panicError wraps a recover() catching a panic()
+type panicError struct {
 	PanicValue interface{}
 	StackTrace string
 }
 
-func (pe PanicError) Error() string {
+func (pe panicError) Error() string {
 	return fmt.Sprintf("panic in TEAL Eval: %v\n%s", pe.PanicValue, pe.StackTrace)
 }
 
 var errLogicSigNotSupported = errors.New("LogicSig not supported")
 var errTooManyArgs = errors.New("LogicSig has too many arguments")
 
-// ClearStateBudgetError allows evaluation to signal that the caller should
-// reject the transaction.  Normally, an error in evaluation would not cause a
-// ClearState txn to fail. However, callers fail a txn for ClearStateBudgetError
-// because the transaction has not provided enough budget to let ClearState do
-// its job.
-type ClearStateBudgetError struct {
-	offered int
+// EvalError indicates AVM evaluation failure
+type EvalError struct {
+	Err        error
+	details    string
+	groupIndex int
+	logicsig   bool
 }
 
-func (e ClearStateBudgetError) Error() string {
-	return fmt.Sprintf("Attempted ClearState execution with low OpcodeBudget %d", e.offered)
+// Error satisfies builtin interface `error`
+func (err EvalError) Error() string {
+	var msg string
+	if err.logicsig {
+		msg = fmt.Sprintf("rejected by logic err=%v", err.Err)
+	} else {
+		msg = fmt.Sprintf("logic eval error: %v", err.Err)
```

Comment:
tricky distinction here between "logic eval error" and "logicsig logic error"

Reply:
I don't like that language necessarily, I'm just matching the existing wording to decrease test churn.

willing to change.